sql/src/main/java/org/apache/apex/malhar/sql/planner/RelNodeVisitor.java

```diff
@@ -0,0 +1,457 @@
+        return Date.class;
+      case TIMESTAMP:
+        return Date.class;
+      case CHAR:
+      case VARCHAR:
+        return String.class;
+      case BINARY:
+      case VARBINARY:
+        return Byte.class;
+      case ANY:
+      case SYMBOL:
+        return Object.class;
+      default:
+        throw new RuntimeException(String.format("Unsupported type %s", type.getSqlTypeName()));
+    }
+  }
+
+  public final RelInfo traverse(RelNode n) throws Exception
+  {
+    List<RelInfo> inputStreams = new ArrayList<>();
+    for (RelNode input : n.getInputs()) {
+      inputStreams.add(traverse(input));
+    }
+
+    RelInfo currentNodeRelInfo;
+
+    if (n instanceof Aggregate) {
+      currentNodeRelInfo = visitAggregate((Aggregate)n, inputStreams);
+    } else if (n instanceof Calc) {
+      currentNodeRelInfo = visitCalc((Calc)n, inputStreams);
+    } else if (n instanceof Collect) {
+      currentNodeRelInfo = visitCollect((Collect)n, inputStreams);
+    } else if (n instanceof Correlate) {
+      currentNodeRelInfo = visitCorrelate((Correlate)n, inputStreams);
+    } else if (n instanceof Delta) {
+      currentNodeRelInfo = visitDelta((Delta)n, inputStreams);
+    } else if (n instanceof Exchange) {
+      currentNodeRelInfo = visitExchange((Exchange)n, inputStreams);
+    } else if (n instanceof Project) {
+      currentNodeRelInfo = visitProject((Project)n, inputStreams);
+    } else if (n instanceof Filter) {
+      currentNodeRelInfo = visitFilter((Filter)n, inputStreams);
+    } else if (n instanceof Sample) {
+      currentNodeRelInfo = visitSample((Sample)n, inputStreams);
+    } else if (n instanceof Sort) {
+      currentNodeRelInfo = visitSort((Sort)n, inputStreams);
+    } else if (n instanceof TableModify) {
+      currentNodeRelInfo = visitTableModify((TableModify)n, inputStreams);
+    } else if (n instanceof TableScan) {
+      currentNodeRelInfo = visitTableScan((TableScan)n, inputStreams);
+    } else if (n instanceof Uncollect) {
+      currentNodeRelInfo = visitUncollect((Uncollect)n, inputStreams);
+    } else if (n instanceof Window) {
+      currentNodeRelInfo = visitWindow((Window)n, inputStreams);
+    } else if (n instanceof Join) {
+      currentNodeRelInfo = visitJoin((Join)n, inputStreams);
+    } else {
+      currentNodeRelInfo = defaultValue(n, inputStreams);
+    }
+
+    if (currentNodeRelInfo != null && inputStreams.size() != 0) {
+      for (int i = 0; i < inputStreams.size(); i++) {
+        RelInfo inputStream = inputStreams.get(i);
+        Operator.OutputPort outputPort = inputStream.getOutPort();
+        Operator.InputPort inputPort = currentNodeRelInfo.getInputPorts().get(i);
+
+        String streamName = OperatorUtils.getUniqueStreamName(inputStream.getRelName(),
+            currentNodeRelInfo.getRelName());
+        Class schema;
+        if (inputStream.getOutRelDataType() != null) {
+          schema = getSchemaForRelDataType(streamName, inputStream.getOutRelDataType());
```

Comment:
Its very much possible. I would say that's an optimization which can be take up later.